internal/cmd/commands/dev/dev.go

```diff
@@ -824,73 +823,84 @@ func (c *Command) Run(args []string) int {
 	c.opsServer = opsServer
 	c.opsServer.Start()
 
-	// Wait for shutdown
-	shutdownTriggered := false
-
-	// Add a force-shutdown goroutine to consume more interrupts
-	abortForceShutdownCh := make(chan struct{})
-	defer close(abortForceShutdownCh)
+	// On first interrupt, enter graceful shutdown for controllers and workers (workers wait for connections to drain)
+	// On second interrupt, force controller shutdown and send workers into shutdown (stop all connections)
+	// On third+ interrupt, force worker shutdown
+	shutdownDone := false
+	shutdownTriggerCount := uint32(0)
 
 	runShutdownLogic := func() {
-		go func() {
-			shutdownCh := make(chan os.Signal, 4)
-			signal.Notify(shutdownCh, os.Interrupt, syscall.SIGTERM)
-			for {
-				select {
-				case <-shutdownCh:
-					c.UI.Error("Forcing shutdown")
-					os.Exit(base.CommandUserError)
+		if sync.LoadUint32(&shutdownTriggerCount) == 1 {
+			go func() {
+				if c.Config.Controller != nil {
+					c.opsServer.WaitIfHealthExists(c.Config.Controller.GracefulShutdownWaitDuration, c.UI)
+				}
+
+				if !c.flagControllerOnly {
+					if !c.flagWorkerAuthStorageSkipCleanup && c.worker.WorkerAuthStorage != nil {
+						c.worker.WorkerAuthStorage.Cleanup()
+					}
+					if err := c.worker.GracefulShutdown(); err != nil {
+						c.UI.Error(fmt.Errorf("Error shutting down worker gracefully: %w", err).Error())
+					}
+					if err := c.worker.Shutdown(); err != nil {
+						c.UI.Error(fmt.Errorf("Error shutting down worker: %w", err).Error())
+					}
+				}
 
-				case <-c.ServerSideShutdownCh:
-					// Drain connections in case this is hit more than once
+				if err := c.controller.Shutdown(); err != nil {
+					c.UI.Error(fmt.Errorf("Error shutting down controller: %w", err).Error())
+				}
 
-				case <-abortForceShutdownCh:
-					// No-op, we just use this to shut down the goroutine
-					return
+				err := c.opsServer.Shutdown()
+				if err != nil {
+					c.UI.Error(fmt.Errorf("Failed to shutdown ops listeners: %w", err).Error())
 				}
-			}
-		}()
+				shutdownDone = true
+			}()
+		}
+
+		if sync.LoadUint32(&shutdownTriggerCount) == 2 && !c.flagControllerOnly {
```

Comment:
These if conditions would probably be a bit cleaner by assigning the value of load to a local var at the start of the function and then using a `switch` statement.

Reply:
Thanks!